[torch] batchnorm: replace control-flow with torch._assert()

stephenyan1231 opened the pull request:
Summary:
For QAT model training, we can use API **prepare_qat_fx()** to convert a plain torch.nn.module into a QAT model of type GraphModule using FX graph mode . Under the hood, it use a symbolic tracer to trace the model and build a model graph.

https://www.internalfb.com/code/fbsource/[cc77b709597e]/fbcode/caffe2/torch/ao/quantization/quantize_fx.py?lines=393

However currently, symbolic tracing does not support control-flow well. We find when we prepare QAT model for Mask2Former 2D segmentation model, **prepare_qat_fx()** will fail due to control flow in 2d batch norm with error below

```
torch.fx.proxy.TraceError: symbolically traced variables cannot be used as inputs to control flow
```


Therefore, we place control-flow with **torch,_assert()** which is friendly to FX graph model and symbolic tracing.

Test Plan:
- Launch a Mask2Former QAT model training locally on devserver: https://www.internalfb.com/code/fbsource/[cc77b709597e]/fbcode/mobile-vision/d2go/projects/scenes/scripts/launch_train_mask2former.sh?lines=13

```
mobile-vision/d2go/projects/scenes/scripts/launch_train_mask2former.sh
```

Differential Revision: D54578538

jbschlosser commented on torch/nn/modules/batchnorm.py:
```diff
@@ -306,10 +306,9 @@ class BatchNorm1d(_BatchNorm):
-                f"expected 2D or 3D input (got {input.dim()}D input)"
-            )
+        torch._assert(
+            input.dim() == 2 or input.dim() == 3, f"expected 2D or 3D input (got {input.dim()}D input)"
```
I agree - seems like we need a symbolically-traceable way to raise a ValueError

stephenyan1231 replied:
@albanD @jbschlosser Do we have such impl of raising a ValueError while also working well with symbolically-traceable? If not, how can I unblock myself?

albanD replied:
Yes they're defined just next to _assert here: https://github.com/pytorch/pytorch/blob/d1510e01faceb7de9095caff75401955ab323310/torch/__init__.py#L1123-L1221

jbschlosser commented:
@stephenyan1231 FYI we can't land with the test failures

stephenyan1231 committed:
[torch] batchnorm: replace control-flow with torch._check_value() (#121391)

Summary:

For QAT model training, we can use API **prepare_qat_fx()** to convert a plain torch.nn.module into a QAT model of type GraphModule using FX graph mode . Under the hood, it use a symbolic tracer to trace the model and build a model graph.

https://www.internalfb.com/code/fbsource/[cc77b709597e]/fbcode/caffe2/torch/ao/quantization/quantize_fx.py?lines=393

However currently, symbolic tracing does not support control-flow well. We find when we prepare QAT model for Mask2Former 2D segmentation model, **prepare_qat_fx()** will fail due to control flow in 2d batch norm with error below

```
torch.fx.proxy.TraceError: symbolically traced variables cannot be used as inputs to control flow
```


Therefore, we place control-flow with **torch,_check_value()** which is friendly to FX graph model and symbolic tracing.

Test Plan:
- Launch a Mask2Former QAT model training locally on devserver: https://www.internalfb.com/code/fbsource/[cc77b709597e]/fbcode/mobile-vision/d2go/projects/scenes/scripts/launch_train_mask2former.sh?lines=13

```
mobile-vision/d2go/projects/scenes/scripts/launch_train_mask2former.sh
```

Differential Revision: D54578538
